### ISSUE TITLE
Unable to refine a datatype

### ISSUE THREAD

**Issue by vikraman** [ucsd-progsys/liquidhaskell#820]
When trying to refine a simple record type like this,
```haskell
{-@ LIQUID "--totality"        @-}

module Data.Foo where

{-@ data Foo = Foo { x :: Int -> Int , y :: Int }
@-}
data Foo = Foo { x :: Int -> Int , y :: Int }
```
```
src/Data/Foo.hs:5:10-13: Error: Specified Type Does Not Refine Haskell Type for Data.Foo.x

 5 | {-@ data Foo = Foo { x :: Int -> Int , y :: Int }
              ^^^^

 The Liquid type

     Data.Foo.Foo -> GHC.Types.Int

 is inconsistent with the Haskell type

     Data.Foo.Foo -> GHC.Types.Int -> GHC.Types.Int
```
Tested with liquidhaskell at commit 349997bec22d4e16e06d6abf17745c0fcd19a073.

**Comment by ranjitjhala** [comment]
Yikes, this looks *bad* -- changing the `Int -> Int` to a type alias
doesn't seem to solve it either. @nikivazou can you take a look?

**Comment by ranjitjhala** [comment]
Btw, if you're super blocked, the below works (but its gross...)

```
{-@ LIQUID "--totality"        @-}

module Data.Foo where

data IFun = IF (Int -> Int)

{-@ data Foo = Foo { x :: IFun
                   , y :: Int }
  @-}
data Foo = Foo { x :: IFun , y :: Int }
```

**Comment by nikivazou** [comment]
Oups, another work around is to refine the data constructors instead of refining the data definition, eg: 
```
{-@ LIQUID "--totality"        @-}
data Foo = Foo { x :: Int -> Int , y :: Int }

{-@ Foo :: (Int -> Int) -> Int -> Foo @-}
{-@ x :: Foo -> Int -> Int @-}
{-@ y :: Foo -> Int @-}
```

Note that the refinement of the data definition directly translates to refinements of data constructors. Though, this specification is not preferred because then you would need to annotate the above specification of `Foo` with all the defined measures. 

You can use either this or @ranjitjhala's alternative until either of us returns from vocations and properly fix this.

**Comment by vikraman** [comment]
Here's another example:
```haskell
{-@ data VerifiedOrd a = VerifiedOrd {
      leq :: (a -> a -> Bool)
    , total :: (x:a -> y:a -> { Prop (leq x y) || Prop (leq y x) })
    , antisym :: (x:a -> y:a -> { Prop (leq x y) || Prop (leq y x) ==> x == y })
    , trans :: (x:a -> y:a -> z:a -> { Prop (leq x y) && Prop (leq y z) ==> Prop (leq x z) })
    , verifiedEq :: VerifiedEq a
    }
@-}
data VerifiedOrd a = VerifiedOrd {
    leq :: a -> a -> Bool
  , total :: a -> a -> Proof
  , antisym :: a -> a -> Proof
  , trans :: a -> a -> a -> Proof
  , verifiedEq :: VerifiedEq a
  }
```
```
.../verified-instances/src/Data/VerifiedOrd.hs:17:10-21: Error: Specified Type Does Not Refine Haskell Type for Data.VerifiedOrd.leq

 17 | {-@ data VerifiedOrd a = VerifiedOrd {
               ^^^^^^^^^^^^

 The Liquid type

     forall a.
Data.VerifiedOrd.VerifiedOrd a -> Data.VerifiedEq.VerifiedEq a

 is inconsistent with the Haskell type

     forall a.
Data.VerifiedOrd.VerifiedOrd a -> a -> a -> GHC.Types.Bool

 defined at .../verified-instances/src/Data/VerifiedOrd.hs:26:5-7

```
But this works:
```haskell
{-@ VerifiedOrd :: leq: (a -> a -> Bool)
                -> total: (x:a -> y:a -> { Prop (leq x y) || Prop (leq y x) })
                -> antisym: (x:a -> y:a -> { Prop (leq x y) || Prop (leq y x) ==> x == y })
                -> trans: (x:a -> y:a -> z:a -> { Prop (leq x y) && Prop (leq y z) ==> Prop (leq x z) })
                -> verifiedEq: VerifiedEq a
                -> VerifiedOrd a
@-}
```

**Comment by nikivazou** [comment]
Fixed by #839
